File: frontend/lib/models/catequisando.dart
import 'pessoa.dart';

enum Genero {
  masculino,
  feminino;

  String get valor => this == Genero.masculino ? 'masculino' : 'feminino';
  String get rotulo => this == Genero.masculino ? 'Masculino' : 'Feminino';

  static Genero? fromValor(String? v) {
    if (v == null) return null;
    return v == 'masculino' ? Genero.masculino : Genero.feminino;
  }
}

enum SituacaoCatequisando {
  ativo,
  crismado,
  transferido;

  String get valor {
    switch (this) {
      case SituacaoCatequisando.ativo:
        return 'ativo';
      case SituacaoCatequisando.crismado:
        return 'crismado';
      case SituacaoCatequisando.transferido:
        return 'transferido';
    }
  }

  String get rotulo {
    switch (this) {
      case SituacaoCatequisando.ativo:
        return 'Ativo';
      case SituacaoCatequisando.crismado:
        return 'Crismado';
      case SituacaoCatequisando.transferido:
        return 'Transferido';
    }
  }

  static SituacaoCatequisando fromValor(String? v) {
    switch (v) {
      case 'crismado':
        return SituacaoCatequisando.crismado;
      case 'transferido':
        return SituacaoCatequisando.transferido;
      default:
        return SituacaoCatequisando.ativo;
    }
  }
}

class Catequisando {
  final String id;
  final String nome;
  final Genero? genero;
  final DateTime? dataNascimento;
  final String? naturalDe;
  final String? estadoCivil;
  final String? profissao;
  final String? residencia;
  final String faseId;
  final String faseNome;
  final String? sectorId;
  final String? sectorNome;
  final SituacaoCatequisando situacao;
  final DateTime? dataSituacao;
  final String? encarregadoNome;
  final String? encarregadoContacto;
  final String? encarregadoParentesco;
  final String? observacoes;

  // Ficha do Catecúmeno — Ano/Número de registo
  final int? fichaAno;
  final String? fichaNumero;

  // Filiação — por referência a Pessoa (reutilizável entre irmãos)
  final String? paiId;
  final Pessoa? pai;
  final String? maeId;
  final Pessoa? mae;
  final String? avoPaternoNome;
  final String? avoPaternaNome;
  final String? avoMaternoNome;
  final String? avoMaternaNome;

  // Padrinhos — idem, por referência (reutilizável entre afilhados)
  final String? padrinhoId;
  final Pessoa? padrinho;
  final String? madrinhaId;
  final Pessoa? madrinha;

  // Eleição e Escrutínios (RCIA) — opcional
  final DateTime? eleicaoData;
  final DateTime? escrutinio1Data;
  final DateTime? escrutinio2Data;
  final DateTime? escrutinio3Data;

  // Baptismo
  final String? baptismoLocal;
  final DateTime? baptismoData;
  final String? baptismoCertidaoUrl;

  // Núcleo (como um Sector, mas do Ministério da Animação dos Núcleos)
  final String? nucleoId;
  final String? nucleoNome;
  final String? nucleoComunidade;

  // Ficha de Baptismo — campos adicionais
  final String? baptismoArquidiocese;
  final String? baptismoLivro;
  final String? baptismoNumeroRegisto;
  final String? baptismoPagina;

  // Ficha de 1ª Comunhão
  final DateTime? primeiraComunhaoData;
  final String? primeiraComunhaoHora;
  final String? primeiraComunhaoLivro;
  final String? primeiraComunhaoNumeroRegisto;
  final String? primeiraComunhaoPagina;

  // Ficha de Crisma
  final DateTime? crismaData;
  final String? crismaHora;
  final String? crismaLivro;
  final String? crismaNumeroRegisto;
  final String? crismaPagina;

  // Transferência para outra paróquia/comunidade
  final String? transferenciaNumero;
  final int? transferenciaAnoGuia;
  final int? transferenciaAnoFrequentado;
  final String? transferenciaDestinoComunidade;
  final String? transferenciaDestinoArquidiocese;
  final String? transferenciaObservacao;
  final DateTime? transferenciaData;

  Catequisando({
    required this.id,
    required this.nome,
    this.genero,
    this.dataNascimento,
    this.naturalDe,
    this.estadoCivil,
    this.profissao,
    this.residencia,
    required this.faseId,
    required this.faseNome,
    this.sectorId,
    this.sectorNome,
    this.situacao = SituacaoCatequisando.ativo,
    this.dataSituacao,
    this.encarregadoNome,
    this.encarregadoContacto,
    this.encarregadoParentesco,
    this.observacoes,
    this.fichaAno,
    this.fichaNumero,
    this.paiId,
    this.pai,
    this.maeId,
    this.mae,
    this.avoPaternoNome,
    this.avoPaternaNome,
    this.avoMaternoNome,
    this.avoMaternaNome,
    this.padrinhoId,
    this.padrinho,
    this.madrinhaId,
    this.madrinha,
    this.eleicaoData,
    this.escrutinio1Data,
    this.escrutinio2Data,
    this.escrutinio3Data,
    this.baptismoLocal,
    this.baptismoData,
    this.baptismoCertidaoUrl,
    this.nucleoId,
    this.nucleoNome,
    this.nucleoComunidade,
    this.baptismoArquidiocese,
    this.baptismoLivro,
    this.baptismoNumeroRegisto,
    this.baptismoPagina,
    this.primeiraComunhaoData,
    this.primeiraComunhaoHora,
    this.primeiraComunhaoLivro,
    this.primeiraComunhaoNumeroRegisto,
    this.primeiraComunhaoPagina,
    this.crismaData,
    this.crismaHora,
    this.crismaLivro,
    this.crismaNumeroRegisto,
    this.crismaPagina,
    this.transferenciaNumero,
    this.transferenciaAnoGuia,
    this.transferenciaAnoFrequentado,
    this.transferenciaDestinoComunidade,
    this.transferenciaDestinoArquidiocese,
    this.transferenciaObservacao,
    this.transferenciaData,
  });

  static DateTime? _data(dynamic v) => v != null ? DateTime.parse(v as String) : null;

  factory Catequisando.fromJson(Map<String, dynamic> json) => Catequisando(
        id: json['id'] as String,
        nome: json['nome'] as String,
        genero: Genero.fromValor(json['genero'] as String?),
        dataNascimento: _data(json['data_nascimento']),
        naturalDe: json['natural_de'] as String?,
        estadoCivil: json['estado_civil'] as String?,
        profissao: json['profissao'] as String?,
        residencia: json['residencia'] as String?,
        faseId: json['fase_id'] as String,
        faseNome: json['fase_nome'] as String,
        sectorId: json['sector_id'] as String?,
        sectorNome: json['sector_nome'] as String?,
        situacao: SituacaoCatequisando.fromValor(json['situacao'] as String?),
        dataSituacao: _data(json['data_situacao']),
        encarregadoNome: json['encarregado_nome'] as String?,
        encarregadoContacto: json['encarregado_contacto'] as String?,
        encarregadoParentesco: json['encarregado_parentesco'] as String?,
        observacoes: json['observacoes'] as String?,
        fichaAno: json['ficha_ano'] as int?,
        fichaNumero: json['ficha_numero'] as String?,
        paiId: json['pai_id'] as String?,
        pai: json['pai'] != null ? Pessoa.fromJson(json['pai'] as Map<String, dynamic>) : null,
        maeId: json['mae_id'] as String?,
        mae: json['mae'] != null ? Pessoa.fromJson(json['mae'] as Map<String, dynamic>) : null,
        avoPaternoNome: json['avo_paterno_nome'] as String?,
        avoPaternaNome: json['avo_paterna_nome'] as String?,
        avoMaternoNome: json['avo_materno_nome'] as String?,
        avoMaternaNome: json['avo_materna_nome'] as String?,
        padrinhoId: json['padrinho_id'] as String?,
        padrinho: json['padrinho'] != null ? Pessoa.fromJson(json['padrinho'] as Map<String, dynamic>) : null,
        madrinhaId: json['madrinha_id'] as String?,
        madrinha: json['madrinha'] != null ? Pessoa.fromJson(json['madrinha'] as Map<String, dynamic>) : null,
        eleicaoData: _data(json['eleicao_data']),
        escrutinio1Data: _data(json['escrutinio_1_data']),
        escrutinio2Data: _data(json['escrutinio_2_data']),
        escrutinio3Data: _data(json['escrutinio_3_data']),
        baptismoLocal: json['baptismo_local'] as String?,
        baptismoData: _data(json['baptismo_data']),
        baptismoCertidaoUrl: json['baptismo_certidao_url'] as String?,
        nucleoId: json['nucleo_id'] as String?,
        nucleoNome: json['nucleo_nome'] as String?,
        nucleoComunidade: json['nucleo_comunidade'] as String?,
        baptismoArquidiocese: json['baptismo_arquidiocese'] as String?,
        baptismoLivro: json['baptismo_livro'] as String?,
        baptismoNumeroRegisto: json['baptismo_numero_registo'] as String?,
        baptismoPagina: json['baptismo_pagina'] as String?,
        primeiraComunhaoData: _data(json['primeira_comunhao_data']),
        primeiraComunhaoHora: json['primeira_comunhao_hora'] as String?,
        primeiraComunhaoLivro: json['primeira_comunhao_livro'] as String?,
        primeiraComunhaoNumeroRegisto: json['primeira_comunhao_numero_registo'] as String?,
        primeiraComunhaoPagina: json['primeira_comunhao_pagina'] as String?,
        crismaData: _data(json['crisma_data']),
        crismaHora: json['crisma_hora'] as String?,
        crismaLivro: json['crisma_livro'] as String?,
        crismaNumeroRegisto: json['crisma_numero_registo'] as String?,
        crismaPagina: json['crisma_pagina'] as String?,
        transferenciaNumero: json['transferencia_numero'] as String?,
        transferenciaAnoGuia: json['transferencia_ano_guia'] as int?,
        transferenciaAnoFrequentado: json['transferencia_ano_frequentado'] as int?,
        transferenciaDestinoComunidade: json['transferencia_destino_comunidade'] as String?,
        transferenciaDestinoArquidiocese: json['transferencia_destino_arquidiocese'] as String?,
        transferenciaObservacao: json['transferencia_observacao'] as String?,
        transferenciaData: _data(json['transferencia_data']),
      );
}

class ErroImportacao {
  final int linha;
  final String motivo;

  ErroImportacao({required this.linha, required this.motivo});

  factory ErroImportacao.fromJson(Map<String, dynamic> json) =>
      ErroImportacao(linha: json['linha'] as int, motivo: json['motivo'] as String);
}

class ImportacaoResultado {
  final int totalLinhas;
  final int criados;
  final List<ErroImportacao> erros;

  ImportacaoResultado({required this.totalLinhas, required this.criados, required this.erros});

  factory ImportacaoResultado.fromJson(Map<String, dynamic> json) => ImportacaoResultado(
        totalLinhas: json['total_linhas'] as int,
        criados: json['criados'] as int,
        erros: (json['erros'] as List).map((e) => ErroImportacao.fromJson(e as Map<String, dynamic>)).toList(),
      );
}

File: backend/app/models/catequisando.py
from datetime import date, datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field

from app.models.pessoa import PessoaOut


class Genero(str, Enum):
    MASCULINO = "masculino"
    FEMININO = "feminino"


class SituacaoCatequisando(str, Enum):
    ATIVO = "ativo"
    CRISMADO = "crismado"
    TRANSFERIDO = "transferido"


class CatequisandoCreate(BaseModel):
    nome: str = Field(..., min_length=2, max_length=150)
    genero: Optional[Genero] = None
    data_nascimento: Optional[date] = None
    natural_de: Optional[str] = Field(None, max_length=100)
    estado_civil: Optional[str] = Field(None, max_length=50)
    profissao: Optional[str] = Field(None, max_length=100)
    residencia: Optional[str] = Field(None, max_length=200)
    fase_id: str
    sector_id: Optional[str] = Field(None, description="Sector pastoral a que também pertence, ex: Acólitos")
    encarregado_nome: Optional[str] = Field(None, max_length=150)
    encarregado_contacto: Optional[str] = Field(None, max_length=50)
    encarregado_parentesco: Optional[str] = Field(None, max_length=50)
    observacoes: Optional[str] = Field(None, max_length=500)

    # Ficha do Catecúmeno — Ano/Número de registo (ex: da ficha em papel)
    ficha_ano: Optional[int] = Field(None, ge=1900, le=2200)
    ficha_numero: Optional[str] = Field(None, max_length=10, description="Pode ter até 5 dígitos")

    # Filiação — referências a pessoas reutilizáveis (ver GET/POST /pessoas),
    # para não duplicar dados entre irmãos ou afilhados da mesma pessoa
    pai_id: Optional[str] = None
    mae_id: Optional[str] = None
    avo_paterno_nome: Optional[str] = Field(None, max_length=150)
    avo_paterna_nome: Optional[str] = Field(None, max_length=150)
    avo_materno_nome: Optional[str] = Field(None, max_length=150)
    avo_materna_nome: Optional[str] = Field(None, max_length=150)

    # Padrinhos — idem, referências reutilizáveis
    padrinho_id: Optional[str] = None
    madrinha_id: Optional[str] = None

    # Eleição e Escrutínios (RCIA) — opcional, só para algumas fases
    eleicao_data: Optional[date] = None
    escrutinio_1_data: Optional[date] = None
    escrutinio_2_data: Optional[date] = None
    escrutinio_3_data: Optional[date] = None

    # Baptismo
    baptismo_local: Optional[str] = Field(None, max_length=150)
    baptismo_data: Optional[date] = None
    baptismo_certidao_url: Optional[str] = Field(None, max_length=500, description="Link do Google Drive")

    # Núcleo — como um Sector, mas do Ministério da Animação dos Núcleos
    nucleo_id: Optional[str] = None
    nucleo_comunidade: Optional[str] = Field(None, max_length=150, description="'da comunidade ___' nas fichas de sacramento")

    # Baptismo — campos adicionais para a Ficha de Baptismo completa
    baptismo_arquidiocese: Optional[str] = Field(None, max_length=150)
    baptismo_livro: Optional[str] = Field(None, max_length=30)
    baptismo_numero_registo: Optional[str] = Field(None, max_length=20)
    baptismo_pagina: Optional[str] = Field(None, max_length=20)

    # Primeira Comunhão
    primeira_comunhao_data: Optional[date] = None
    primeira_comunhao_hora: Optional[str] = Field(None, max_length=20)
    primeira_comunhao_livro: Optional[str] = Field(None, max_length=30)
    primeira_comunhao_numero_registo: Optional[str] = Field(None, max_length=20)
    primeira_comunhao_pagina: Optional[str] = Field(None, max_length=20)

    # Crisma
    crisma_data: Optional[date] = None
    crisma_hora: Optional[str] = Field(None, max_length=20)
    crisma_livro: Optional[str] = Field(None, max_length=30)
    crisma_numero_registo: Optional[str] = Field(None, max_length=20)
    crisma_pagina: Optional[str] = Field(None, max_length=20)


class CatequisandoUpdate(BaseModel):
    nome: Optional[str] = Field(None, min_length=2, max_length=150)
    genero: Optional[Genero] = None
    data_nascimento: Optional[date] = None
    natural_de: Optional[str] = Field(None, max_length=100)
    estado_civil: Optional[str] = Field(None, max_length=50)
    profissao: Optional[str] = Field(None, max_length=100)
    residencia: Optional[str] = Field(None, max_length=200)
    fase_id: Optional[str] = None
    sector_id: Optional[str] = None
    situacao: Optional[SituacaoCatequisando] = None
    encarregado_nome: Optional[str] = Field(None, max_length=150)
    encarregado_contacto: Optional[str] = Field(None, max_length=50)
    encarregado_parentesco: Optional[str] = Field(None, max_length=50)
    observacoes: Optional[str] = Field(None, max_length=500)

    ficha_ano: Optional[int] = Field(None, ge=1900, le=2200)
    ficha_numero: Optional[str] = Field(None, max_length=10)

    pai_id: Optional[str] = None
    mae_id: Optional[str] = None
    avo_paterno_nome: Optional[str] = Field(None, max_length=150)
    avo_paterna_nome: Optional[str] = Field(None, max_length=150)
    avo_materno_nome: Optional[str] = Field(None, max_length=150)
    avo_materna_nome: Optional[str] = Field(None, max_length=150)

    padrinho_id: Optional[str] = None
    madrinha_id: Optional[str] = None

    eleicao_data: Optional[date] = None
    escrutinio_1_data: Optional[date] = None
    escrutinio_2_data: Optional[date] = None
    escrutinio_3_data: Optional[date] = None

    baptismo_local: Optional[str] = Field(None, max_length=150)
    baptismo_data: Optional[date] = None
    baptismo_certidao_url: Optional[str] = Field(None, max_length=500)

    nucleo_id: Optional[str] = None
    nucleo_comunidade: Optional[str] = Field(None, max_length=150)

    baptismo_arquidiocese: Optional[str] = Field(None, max_length=150)
    baptismo_livro: Optional[str] = Field(None, max_length=30)
    baptismo_numero_registo: Optional[str] = Field(None, max_length=20)
    baptismo_pagina: Optional[str] = Field(None, max_length=20)

    primeira_comunhao_data: Optional[date] = None
    primeira_comunhao_hora: Optional[str] = Field(None, max_length=20)
    primeira_comunhao_livro: Optional[str] = Field(None, max_length=30)
    primeira_comunhao_numero_registo: Optional[str] = Field(None, max_length=20)
    primeira_comunhao_pagina: Optional[str] = Field(None, max_length=20)

    crisma_data: Optional[date] = None
    crisma_hora: Optional[str] = Field(None, max_length=20)
    crisma_livro: Optional[str] = Field(None, max_length=30)
    crisma_numero_registo: Optional[str] = Field(None, max_length=20)
    crisma_pagina: Optional[str] = Field(None, max_length=20)


class CatequisandoOut(BaseModel):
    id: str
    nome: str
    genero: Optional[Genero] = None
    data_nascimento: Optional[date] = None
    natural_de: Optional[str] = None
    estado_civil: Optional[str] = None
    profissao: Optional[str] = None
    residencia: Optional[str] = None
    fase_id: str
    fase_nome: str
    sector_id: Optional[str] = None
    sector_nome: Optional[str] = None
    situacao: SituacaoCatequisando = SituacaoCatequisando.ATIVO
    data_situacao: Optional[date] = None
    encarregado_nome: Optional[str] = None
    encarregado_contacto: Optional[str] = None
    encarregado_parentesco: Optional[str] = None
    observacoes: Optional[str] = None

    ficha_ano: Optional[int] = None
    ficha_numero: Optional[str] = None

    pai_id: Optional[str] = None
    pai: Optional[PessoaOut] = None
    mae_id: Optional[str] = None
    mae: Optional[PessoaOut] = None
    avo_paterno_nome: Optional[str] = None
    avo_paterna_nome: Optional[str] = None
    avo_materno_nome: Optional[str] = None
    avo_materna_nome: Optional[str] = None

    padrinho_id: Optional[str] = None
    padrinho: Optional[PessoaOut] = None
    madrinha_id: Optional[str] = None
    madrinha: Optional[PessoaOut] = None

    eleicao_data: Optional[date] = None
    escrutinio_1_data: Optional[date] = None
    escrutinio_2_data: Optional[date] = None
    escrutinio_3_data: Optional[date] = None

    baptismo_local: Optional[str] = None
    baptismo_data: Optional[date] = None
    baptismo_certidao_url: Optional[str] = None

    nucleo_id: Optional[str] = None
    nucleo_nome: Optional[str] = None
    nucleo_comunidade: Optional[str] = None

    baptismo_arquidiocese: Optional[str] = None
    baptismo_livro: Optional[str] = None
    baptismo_numero_registo: Optional[str] = None
    baptismo_pagina: Optional[str] = None

    primeira_comunhao_data: Optional[date] = None
    primeira_comunhao_hora: Optional[str] = None
    primeira_comunhao_livro: Optional[str] = None
    primeira_comunhao_numero_registo: Optional[str] = None
    primeira_comunhao_pagina: Optional[str] = None

    crisma_data: Optional[date] = None
    crisma_hora: Optional[str] = None
    crisma_livro: Optional[str] = None
    crisma_numero_registo: Optional[str] = None
    crisma_pagina: Optional[str] = None

    # Transferência para outra paróquia/comunidade
    transferencia_numero: Optional[str] = None
    transferencia_ano_guia: Optional[int] = None
    transferencia_ano_frequentado: Optional[int] = None
    transferencia_destino_comunidade: Optional[str] = None
    transferencia_destino_arquidiocese: Optional[str] = None
    transferencia_observacao: Optional[str] = None
    transferencia_data: Optional[date] = None

    criado_em: datetime


class TransferenciaRequest(BaseModel):
    numero: str = Field(..., min_length=1, max_length=10)
    ano_guia: int = Field(..., ge=1900, le=2200)
    ano_frequentado: int = Field(..., ge=1900, le=2200)
    destino_comunidade: str = Field(..., min_length=2, max_length=200)
    destino_arquidiocese: str = Field(..., min_length=2, max_length=150)
    observacao: Optional[str] = Field(None, max_length=500)
    data_impressao: Optional[date] = None


class ErroImportacao(BaseModel):
    linha: int
    motivo: str


class ImportacaoResultado(BaseModel):
    total_linhas: int
    criados: int
    erros: list[ErroImportacao]
